### PR TITLE
Reduce L0 test time

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Category:

**Other**

## Description:

- Separate slow tests to reduce L0 test time
- Trim the combination of test arguments (no need cartesian product always)

### Affected modules and functionalities:

Test

### Tests:

- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/test/python/operator_2/test_filter.py [hunk 1]
```diff
@@ -139,6 +139,7 @@ def test_image_pipeline(dtype, batch_size, border, mode):
         check_batch(filtered_imgs, baseline, max_allowed_error=atol)
 
 
+@attr('slow')
```
Comment (jantonguirao) [inline dali/test/python/operator_2/test_filter.py:142]: only relevant for nose (not nose2)

File: dali/test/python/operator_2/test_filter.py [hunk 2]
```diff
@@ -177,7 +178,7 @@ def test_image_pipeline(dtype, batch_size, border, mode):
                          (1, 1, 1024 * 1024)], [(1, 256, 1), (1, 1, 257),
                                                 (258, 1, 1)], 3, "101", "same"),
 )
-def test_samples(dtype, sample_layout, sample_shapes, filter_shapes, batch_size, border, mode):
+def slow_test_samples(dtype, sample_layout, sample_shapes, filter_shapes, batch_size, border, mode):
```
Comment (jantonguirao) [inline dali/test/python/operator_2/test_filter.py:181]: relevant for our nose2 invokation (filters slow_* tests)

File: qa/TL0_python-self-test-operators_1/test_body.sh [hunk 3]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash -e
 
 test_py_with_framework() {
-    ${python_new_invoke_test} -A '!slow' -s operator_1
+    ${python_new_invoke_test} -s operator_1
```
Comment (jantonguirao) [inline qa/TL0_python-self-test-operators_1/test_body.sh:4]: this `-A '!slow'` didn't do anything

**Comment by jantonguirao** [comment]
!build

File: qa/TL0_python-self-test-operators_1/test_body.sh (hunk 3 above)
Comment (JanuszL) [reply qa/TL0_python-self-test-operators_1/test_body.sh:4]: Mhhh
I was able to run this simple example:
```
from nose.plugins.attrib import attr


def test1():
    pass

@attr("slow")
def test2():
    pass
```
with `python3 -m nose2 --verbose -A '!slow' -s test` and `python3 -m nose2 --verbose -A 'slow' -s test` and it provided the expected results.

I think now we just rely on `unittest_slow.cfg` which defines the `slow_` test prefix instead of using the attribute.

File: dali/test/python/operator_2/test_warp.py [hunk 4]
```diff
@@ -302,8 +303,8 @@ def get_data():
     for c in range(channels):
         assert out[0, 0, c] == c
 
-
-def test_extremely_large_data():
+@attr('slow')
+def slow_test_extremely_large_data():
```
Comment (mzient) [inline dali/test/python/operator_2/test_warp.py:307]: Strictly speaking, this test isn't slow. It takes 2s on A100 - unless you hit a slow node and then it can take 300s, but I don't think that getting rid of all tests that may hurt some machine's feelings is a good idea.

File: dali/test/python/operator_2/test_reduce.py [hunk 5]
```diff
@@ -493,8 +493,8 @@ def _test_std_dev_large_data(rank, axes, device):
             ref = np.std(batch[i].astype(np.float64), axis=axes, ddof=0)
             assert np.allclose(out[i], ref, 1e-5, 1e-5)
 
-
-def test_std_dev_large_data():
+attr('slow')
+def slow_test_std_dev_large_data():
```
Comment (mzient) [inline dali/test/python/operator_2/test_reduce.py:497]: Admittedly, this test is quite slow (18s on my machine), however, I'd keep one of the tests with "large data" - preferably this one. It's actually the only test that can quite reliably test batch subdivision mechanism and distribution of samples to different kernels. Note that it always uses one reduction (`sum`). Also, this test checks the numerical stability of summation and for that we need many numbers.

Comment (mzient) [reply dali/test/python/operator_2/test_reduce.py:497]: BTW - I was able to cut the test time from 18s to 11s (for test with stddev) and from 12s to 8s (without stddev) by speeding up data generation:
```Python
def fast_large_random_batches(rank, batch_size, num_batches, lo=0, hi=1):
    max_extent = min(65536, int(np.floor(10000000**(1/rank))))

    # generate a maximum size buffer pre-filled with random numbers
    buf = np.random.uniform(low=lo, high=hi, size=max_extent**rank).astype(np.float32)

    data = []
    for _ in range(num_batches):
        batch = []
        for _ in range(batch_size):
            size = np.random.randint(1, max_extent, size=rank)
            vol = np.prod(size)
            # now that we know the actual volume of the sample, we can pick a random
            # location in the pre-filled buffer
            offset = np.random.randint(0, (buf.size - vol) + 1)
            # take a slice and reshape it to the desired shape - these are constant time operations
            sample = buf[offset:offset+vol].reshape(size)
            batch.append(sample)
        data.append(batch)
    return data
```

and the usage:
```Python
@nottest
def _test_reduce_large_data(rank, axes, device):
    batch_size = 16
    num_batches = 2
    data = fast_large_random_batches(rank, batch_size, num_batches)
```
and
```Python
@nottest
def _test_std_dev_large_data(rank, axes, device):
    batch_size = 16
    num_batches = 2
    data = fast_large_random_batches(rank, batch_size, num_batches, 1, 2);
```

File: qa/TL0_python-self-test-operators_1/test_body.sh (hunk 3 above)
Comment (JanuszL) [reply qa/TL0_python-self-test-operators_1/test_body.sh:4]: Ok, it doesn't work with `yield`, only with `params`. I'm fine either way.

File: dali/test/python/operator_2/test_reduce.py [hunk 6]
```diff
@@ -229,8 +228,8 @@ def test_reduce():
     for keep_dims in [False, True]:
         for reduction_name in reductions:
             for batch_gen in batch_gens:
-                for type_id in types:
-                    yield run_reduce, keep_dims, reduction_name, batch_gen, type_id
+                type_id = np.random.choice(types)
```
Comment (jantonguirao) [inline dali/test/python/operator_2/test_reduce.py:231]: all combinations of inputs is already tested in many more tests (the ones with overflow, type promotion, etc)

File: dali/test/python/operator_2/test_reduce.py [hunk 7]
```diff
@@ -294,14 +293,14 @@ def test_sum_with_output_type():
...
-                for type_map in types:
-                    input_type = type_map[0]
-                    for output_type in type_map[1]:
-                        yield (run_reduce,
-                               keep_dims, reduction_name, batch_gen, input_type, output_type)
+    for reduction_name in reductions:
+        for batch_gen in batch_gens:
+            for type_map in types:
+                input_type = type_map[0]
+                keep_dims = np.random.choice([False, True])
```
Comment (jantonguirao) [inline dali/test/python/operator_2/test_reduce.py:300]: don't need to duplicate tests again for keep_dims

File: dali/test/python/operator_2/test_rotate.py [hunk 8]
```diff
@@ -234,19 +235,18 @@ def create_pipeline(backend, *args):
 
 
 def run_cases(backend1, backend2, epsilon):
-    for batch_size in [1, 4, 19]:
-        for output_size in [None, (160, 240)]:
-            for (itype, otype) in io_types:
+    for output_size in [None, (160, 240)]:
+        for (itype, otype) in io_types:
+            batch_size = np.random.choice([1, 4, 19])
```
Comment (jantonguirao) [inline dali/test/python/operator_2/test_rotate.py:240]: no need to triplicate all tests for different batch size. Like this we will hit each of them many times anyway

File: dali/test/python/operator_2/test_warp.py [hunk 9]
```diff
@@ -190,79 +190,41 @@ def compare(pipe1, pipe2, max_err):
...
-                gpu_pipeline.build()
-
-                compare(cpu_pipeline, gpu_pipeline, 1.0001)
+    def impl(batch_size, use_input, otype, itype, inv_map):
+        cpu_pipeline = WarpPipeline("cpu", batch_size, otype, itype, use_input, inv_map=inv_map)
+        cpu_pipeline.build()
+        gpu_pipeline = WarpPipeline("gpu", batch_size, otype, itype, use_input, inv_map=inv_map)
+        gpu_pipeline.build()
 
+    random.seed(1005)
```
Comment (mzient) [inline dali/test/python/operator_2/test_warp.py:222]: Make sure to pick a good seed - I tried that and with 1005 the distribution wasn't good.

File: dali/test/python/operator_2/test_reduce.py [hunk 10]
```diff
@@ -427,18 +425,44 @@
...
+    max_vol = 10000000
+    max_extent = min(65536, int(np.floor(max_vol**(1/rank))))
+
+    # generate a maximum size buffer pre-filled with random numbers
+    global _random_buf
+    global _random_lo
+    global _random_hi
+    if (_random_buf is None or _random_buf.size < max_extent**rank or
+        _random_lo != lo or _random_hi != hi):
+        _random_lo = lo
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/test/python/operator_2/test_reduce.py:442]: it is used to remember the last ine (and if it is different, regenerate)

**Review by mzient — APPROVED** [review]

**Review by awolant — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build
